File: go/0496_next_greater_element_I.go
package main

// time: O(n)
// space: O(n)

func nextGreaterElement(nums1 []int, nums2 []int) []int {

	stack := []int{}
	nextGreater := make(map[int]int)

	for _, num := range nums2 {
		// While the stack is not empty and the current number is greater than the top of the stack
		for len(stack) > 0 && num > stack[len(stack)-1] {
			last := stack[len(stack)-1]
			nextGreater[last] = num
			stack = stack[:len(stack)-1]
		}

		stack = append(stack, num)
	}

	// Set missing nextGreater to -1
	for len(stack) > 0 {
		nextGreater[stack[len(stack)-1]] = -1
		stack = stack[:len(stack)-1]
	}

	// map nextGreater to nums1
	for i, v := range nums1 {
		nums1[i] = nextGreater[v]
	}

	return nums1
}

File: csharp/0496-next-greater-element-I.cs
public partial class Solution {
    public int[] NextGreaterElement(int[] nums1, int[] nums2) {
        var stack = new Stack<int>();
        var nextGreater = new Dictionary<int, int>();

        foreach (var num in nums2) {
            while (stack.Count > 0 && num > stack.Peek()) {
                nextGreater[stack.Pop()] = num;
            }
            stack.Push(num);
        }

        foreach (var num in stack) {
            nextGreater.Add(num, -1);
        }

        for (var i = 0; i < nums1.Length; i++) {
            nums1[i] = nextGreater[nums1[i]];
        }

        return nums1;
    }
}



public class Program {
    public static void Main() {
        var nums1 = new int[]{4, 1, 2};
        var nums2 = new int[]{1,3,4,2};

        var s = new Solution();
        var res = s.NextGreaterElement(nums1, nums2);
        foreach (var r in res) {
            Console.WriteLine(r);
        }
    }
}
